gossip/gossip.go

```diff
@@ -251,17 +243,51 @@ func (g *Gossip) AddInfo(key string, val interface{}, ttl time.Duration) error {
 	return err
 }
 
+// AddInfoProto adds or updates an info object. Returns an error if info
+// couldn't be added.
+func (g *Gossip) AddInfoProto(key string, proto gogoproto.Message, ttl time.Duration) error {
+	bytes, err := gogoproto.Marshal(proto)
+	if err != nil {
+		return err
+	}
+	return g.AddInfo(key, bytes, ttl)
+}
+
 // GetInfo returns an info value by key or an error if specified
 // key does not exist or has expired.
-func (g *Gossip) GetInfo(key string) (interface{}, error) {
+func (g *Gossip) GetInfo(key string) ([]byte, error) {
 	g.mu.Lock()
-	defer g.mu.Unlock()
-	if i := g.is.getInfo(key); i != nil {
-		return i.Val, nil
+	i := g.is.getInfo(key)
+	g.mu.Unlock()
```

Comment:
why not hold the lock in case anything else changes under you and defer instead?